# Автоматическая обработка текстов 
## Домашнее задание 1 [20 баллов] до 23:59 28.04.2018

В этом домашнем задании вам потребуется написать генератор описания прогноза погоды на следующую неделю в каком-нибудь городе. Домашнее задание состоит из двух частей:
1. Генерация текста по шаблону
2. Извлечение телефонных номеров из текста.

## Часть 1 [10 баллов]
Задание состоит из следующих пунктов:
1. Скачивание данных о состоянии погоды в городе 
2. Генерация описания прогноза
3. Творческая часть

Все три пункта можно считать независимыми – вы можете сделать один или два из них, однако мы настоятельно советуем выполнить все три. Все инструкции по выполнению домашнего задания – ниже. 

### 1. Сбор данных [3 балла]


Пример: прогноз на 10 ближайших дней в Москве – https://www.gismeteo.ru/weather-moscow-4368/10-days/

Используя известные вам библиотеки для работы с протоколом http и html кодом, извлеките прогноз на ближайшие 10 дней, начиная со дня, когда вы начали делать домашнее задание, с любого сервиса с прогнозом погоды или используя его API.
Примеры сервисов:
* gismeteo.ru, https://www.gismeteo.ru/api/ – Gismeteo, API Gismeteo 
* https://tech.yandex.ru/weather/ – API Yandex.Погоды
* https://sinoptik.com.ru – Sinoptik
* любой другой 

Резльтатом сбора данных должна быть таблица со следующими строками:
* минимальная температура
* максимальная температура
* скорость ветра
* уровень осадков 

В столбцах таблицы должны быть даты и дни недели.  Пример итоговой таблицы вы найдете в следующей части задания. 

In [3]:
import requests

r = requests.get("http://api.wunderground.com/api/f58c032a9510345a/forecast10day/q/Russia/Moscow.json")
print(r.status_code)
response_data = r.json()

200


In [11]:
data = response_data['forecast']['simpleforecast']['forecastday']

In [46]:
from datetime import datetime
import math
import locale

locale.setlocale(locale.LC_ALL, 'ru_RU.UTF-8')

class Weather(object):
    def __init__(self, epoch, low, high, wind, precip):
        self.dt = datetime.fromtimestamp(int(epoch))
        self.low = float(low)
        self.high = float(high)
        self.wind = float(wind)
        self.precip = float(precip)
        
    def __repr__(self):
        return f"{self.dt} temp: {self.low}..{self.high} wind: {self.wind}m/s precip: {self.precip}mm"

forecast = list()
for day in data:
    w = Weather(
        epoch=day['date']['epoch'],
        low=day['low']['celsius'],
        high=day['high']['celsius'],
        wind=math.ceil(day['maxwind']['kph']*1000.0/3600),
        precip=day['qpf_allday']['mm']
    )
    forecast.append(w)
    
forecast


[2018-04-27 19:00:00 temp: 4.0..15.0 wind: 2.0m/s precip: 0.0mm,
 2018-04-28 19:00:00 temp: 5.0..15.0 wind: 9.0m/s precip: 0.0mm,
 2018-04-29 19:00:00 temp: 8.0..16.0 wind: 7.0m/s precip: 0.0mm,
 2018-04-30 19:00:00 temp: 11.0..21.0 wind: 7.0m/s precip: 4.0mm,
 2018-05-01 19:00:00 temp: 11.0..24.0 wind: 9.0m/s precip: 5.0mm,
 2018-05-02 19:00:00 temp: 10.0..19.0 wind: 7.0m/s precip: 0.0mm,
 2018-05-03 19:00:00 temp: 8.0..18.0 wind: 7.0m/s precip: 0.0mm,
 2018-05-04 19:00:00 temp: 9.0..19.0 wind: 7.0m/s precip: 8.0mm,
 2018-05-05 19:00:00 temp: 8.0..18.0 wind: 5.0m/s precip: 1.0mm,
 2018-05-06 19:00:00 temp: 8.0..17.0 wind: 7.0m/s precip: 0.0mm]

In [47]:
import pandas as pd

In [48]:
index = [w.dt.strftime("%d.%m (%a)") for w in forecast]
d = {
    'минимальная температура': [w.low for w in forecast],
    'максимальная температура': [w.high for w in forecast],
    'скорость ветра': [w.wind for w in forecast],
    'уровень осадков': [w.precip for w in forecast]
}

f = pd.DataFrame(d, index=index)

In [49]:
f.T

27.04 (пт)  28.04 (сб)  29.04 (вс)  30.04 (пн)  \
максимальная температура        15.0        15.0        16.0        21.0   
минимальная температура          4.0         5.0         8.0        11.0   
скорость ветра                   2.0         9.0         7.0         7.0   
уровень осадков                  0.0         0.0         0.0         4.0   

                          01.05 (вт)  02.05 (ср)  03.05 (чт)  04.05 (пт)  \
максимальная температура        24.0        19.0        18.0        19.0   
минимальная температура         11.0        10.0         8.0         9.0   
скорость ветра                   9.0         7.0         7.0         7.0   
уровень осадков                  5.0         0.0         0.0         8.0   

                          05.05 (сб)  06.05 (вс)  
максимальная температура        18.0        17.0  
минимальная температура          8.0         8.0  
скорость ветра                   5.0         7.0  
уровень осадков                  1.0         0.0

### 2. Генератор описания прогноза погоды [4 балла]

Если у вас не получилось извлечь прогноз погоды в предыдущей части задания, воспользуйтесь таблицей ниже.
В ней приведен прогноз четырех показателей на первые 10 дней февраля в Москве – минимальная и максимальная температура, скорость ветра и уровень осадков. 

|                | 02.02 (пт) | 03.02 (сб) | 04.02 (вс)| 05.02 (пн) | 06.02 (вт) | 07.02 (пн) | 08.02 (ср) | 09.02 (ср) | 10.02 (сб) | 11.02 (вс)
|----------------|-------|-------|-------|-------|-------|-------|-------|
| минимальная температура    | -9    | -1    | -8    | -13    | -12    | -15    | -21    | -14 |-8 |-8
| максимальная температура    | -1    | +1    | -2    | -9   | -11    | -12    | -16    |-5    |-6    |-5|
| скорость ветра | 10    | 13    | 15    | 15   |11    | 6    | 7 | 9 | 8 |12
| уровень осадков         | 1.35  | 8.6  | 15.5  | 6.6   | 2.7   | 2.1   | 0   | 3.2   |0.8  | 0.4

Прогноз погоды должен состоять из следующих (или подобным им) предложений, генерируемых по шаблонам (ниже три шаблона):
* В день1 похолодает / потеплеет на X градус (-а, -ов) по сравнению с день2
    * *В четверг в НазваниеГорода потеплеет на 7 градусов по сравнению со средой*
* Скорость ветра изменится на X единиц в день1 по сравнению с день2.
    * *Скорость ветра изменится на 3 км/час в понедельник по сравнению с пятницей*
* Уровень осадков повысится / понизится на X единиц за Y дней. 
    * *Уровень осадков понится на 3.85 мм за 7 дней*
    * *Выпадет 10 см снега за ближайшие 7 дней * 


Вместо НазваниеГорода и дней недели подставьте название выбранного вами города и дни недели, используя фунцкии для согласования существительных с предлогами. Используйте функции для согласования числительного с существительным для согласования длительности промежутков времени и слова "день" и чисел  с последующими единицами измерения.

In [37]:
import pymorphy2
morph = pymorphy2.analyzer.MorphAnalyzer()
city = 'Москва'

In [65]:
def forecast_one(city, day1, day2, forecast):
    # В _четверг_ в _Городе_ потеплеет на _7_ градусов по сравнению со _средой_
    
    day1w = forecast[day1].dt.strftime("%A")
    prep1 = 'В' if day1w != 'вторник' else 'Во'
    day1w = morph.parse(day1w)[0].inflect({'accs'}).word
    
    day2w = forecast[day2].dt.strftime("%A")
    prep2 = 'с' if day2w != 'среда' else 'со'
    day2w = morph.parse(day2w)[0].inflect({'ablt'}).word
    
    city = morph.parse(city)[0].inflect({'loct'}).word

    if forecast[day1].high == forecast[day2].high:
        return f'{prep1} {day1w} в {city[0].upper()+city[1:]} погода не изменится по сравнению {prep2} {day2w}.'
    
    change = 'потеплеет' if forecast[day1].high > forecast[day2].high else 'похолодает'
    amount = math.ceil(abs(forecast[day1].high - forecast[day2].high))
    degree = morph.parse('градус')[0].make_agree_with_number(int(amount)).word
    
    return f'{prep1} {day1w} в {city[0].upper()+city[1:]} {change} на {amount} {degree} по сравнению {prep2} {day2w}.'

for i in range(9):
    print(forecast_one('Москва', i+1, i, forecast))

В субботу в Москве погода не изменится по сравнению с пятницей.
В воскресенье в Москве потеплеет на 1 градус по сравнению с субботой.
В понедельник в Москве потеплеет на 5 градусов по сравнению с воскресеньем.
Во вторник в Москве потеплеет на 3 градуса по сравнению с понедельником.
В среду в Москве похолодает на 5 градусов по сравнению с вторником.
В четверг в Москве похолодает на 1 градус по сравнению со средой.
В пятницу в Москве потеплеет на 1 градус по сравнению с четвергом.
В субботу в Москве похолодает на 1 градус по сравнению с пятницей.
В воскресенье в Москве похолодает на 1 градус по сравнению с субботой.


In [68]:
def forecast_two(city, day1, day2, forecast):
    # Скорость ветра изменится на 3 км/час в понедельник по сравнению с пятницей
    
    day1w = forecast[day1].dt.strftime("%A")
    prep1 = 'в' if day1w != 'вторник' else 'Во'
    day1w = morph.parse(day1w)[0].inflect({'accs'}).word
    
    day2w = forecast[day2].dt.strftime("%A")
    prep2 = 'с' if day2w != 'среда' else 'со'
    day2w = morph.parse(day2w)[0].inflect({'ablt'}).word

    if forecast[day1].wind == forecast[day2].wind:
        return f'Скорость ветра не изменится {prep1} {day1w} по сравнению {prep2} {day2w}.'
    
    change = 'увеличится' if forecast[day1].wind > forecast[day2].wind else 'уменьшится'
    amount = math.ceil(abs(forecast[day1].wind - forecast[day2].wind))
    
    return f'Скорость ветра {change} на {amount} м/с {prep1} {day1w} по сравнению {prep2} {day2w}.'

for i in range(9):
    print(forecast_two('Москва', i+1, i, forecast))

Скорость ветра увеличится на 7 м/с в субботу по сравнению с пятницей.
Скорость ветра уменьшится на 2 м/с в воскресенье по сравнению с субботой.
Скорость ветра не изменится в понедельник по сравнению с воскресеньем.
Скорость ветра увеличится на 2 м/с Во вторник по сравнению с понедельником.
Скорость ветра уменьшится на 2 м/с в среду по сравнению с вторником.
Скорость ветра не изменится в четверг по сравнению со средой.
Скорость ветра не изменится в пятницу по сравнению с четвергом.
Скорость ветра уменьшится на 2 м/с в субботу по сравнению с пятницей.
Скорость ветра увеличится на 2 м/с в воскресенье по сравнению с субботой.


In [76]:
def forecast_three(city, day1, day2, forecast):
    # Уровень осадков понизится на 3.85 мм за 7 дней
    
    interval = day1 - day2
    days = morph.parse('дни')[0].make_agree_with_number(interval).word
    
    if forecast[day1].precip == forecast[day2].precip:
        return f'Уровень осадков не изменится за {interval} {days}.'
    
    change = 'повысится' if forecast[day1].wind > forecast[day2].wind else 'понизится'
    amount = abs(forecast[day1].wind - forecast[day2].wind)
    
    return f'Уровень осадков {change} на {amount} мм за {interval} {days}.'

for i in range(1,9):
    print(forecast_three('Москва', i, 0, forecast))

Уровень осадков не изменится за 1 день.
Уровень осадков не изменится за 2 дня.
Уровень осадков повысится на 5.0 мм за 3 дня.
Уровень осадков повысится на 7.0 мм за 4 дня.
Уровень осадков не изменится за 5 дней.
Уровень осадков не изменится за 6 дней.
Уровень осадков повысится на 5.0 мм за 7 дней.
Уровень осадков повысится на 3.0 мм за 8 дней.


### 3. Ответьте на вопросы [3 балла]
* В каких других задачах (помимо описания прогноза погоды) может понадобиться генерировать текст по шаблонам? В каких задачах может понадобиться генерировать текст об изменении числовых показателей по шаблонам?
* Шаблоны, которые вы использовали в этом задании, имеют фиксированную структуру. Фактически, ваша задача заключалась в том, чтобы подставить в шаблон число и согласовать единицы измерения с этим числом или подставить в шаблон название города и согласовать его с предлогом. Как можно разнообразить эти шаблоны? Как знание синтаксической структуры предложения может помочь в этой задаче? 

Генерировать текст по шаблонам может быть необходимо на любом веб-сайте, поисковой выдаче. Например, "по заданному условию найдено X документов(а)"

Текст об изменениях числовых показателей может понадобится для информирования о ходе биржевых торгов, спортивных соревнований и т.п. Везде, где есть поток числовых данных.

Можно использовать несколько шаблонов и чередовать их или выбирать в случайном порядке. Например, "Завтра в Москве похолодает на 3 градуса", "Завтра термометры покажут на 3 градуса меньше", "Завтра столбик термометра опустится на 3 градуса". 

Можно также попробовать менять порядок слов в предложении, для этого может пригодится дерево зависимостей слов.

## Часть 2 [10 баллов]
Данная часть состоит из пяти пунктов.
1. Генерация датасета.
2. Обучение классификатора.
3. Поиск номеров регулярным выражением.
4. Оценка качества работы.

### 1. Генерация датасета [2 балла].
Возьмите любой русский текст (можно взять твиты из семинара). Разбейте его на предложения. Вставьте <b>случайные</b> телефонные номера в любую часть предложения.

Примеры номеров:
>89268659970

>8(495)3451212

>+7(234)456-78-90

>8(956)234-23-23

>85679962312 

>+7-845-344-23-65

Можете придумать еще вид номеров. Чтобы сгенерировать случайный номер используйте Шаблон и вставляйте в него случайный номер. Запомните позицию вставки (номер слова, после которого будет вставлен номер).

Сначала научимся делать шаблоны отображения телефонного номера, можно использовать метод string.format()

In [78]:
"+7-{}{}{}-{}{}{}-{}{}{}{}".format(*"5679962312")

'+7-567-996-2312'

Напишем функцию генерации случайного номера по случайному шаблону:

In [98]:
import random

phone_templates = [
    "+7-{}{}{}-{}{}{}-{}{}{}{}",
    "+7-({}{}{})-{}{}{}-{}{}{}{}",
    "+7-{}{}{}-{}{}{}-{}{}-{}{}",
    "8{}{}{}{}{}{}{}{}{}{}",
    "8({}{}{}){}{}{}{}{}{}{}",
    "8({}{}{})-{}{}{}-{}{}-{}{}"
]

def generate_phone():
    phone = ''.join(map(str, random.choices(range(10), k=10)))
    template = random.choice(phone_templates)
    return template.format(*phone)
    
for i in range(10):
    print(generate_phone())

8(348)6723345
8(178)-741-15-09
+7-448-067-9861
+7-014-658-98-99
+7-(902)-652-3576
8(501)3663354
+7-973-117-60-30
+7-052-037-2567
+7-(962)-700-7442
+7-718-093-2663


Прочитаем текст и разобъем его на предложения:

In [95]:
with open('kapitan.txt', 'r', encoding='utf8') as f:
    text_plain = f.read()

In [96]:
import nltk

def split_data(text):
    res = [line.replace("\n", "") for line in nltk.sent_tokenize(text) if len(line.replace("\n", ""))]
    return res

text = split_data(text_plain)

Посмотрим характеристики полученного набора предложений:

In [97]:
import pandas as pd

df = pd.DataFrame(text)
df[0][10:20]

10                       Нас было девять человек детей.
11      Все мои братья и сестры умерли во младенчестве.
12    Матушка была еще мною брюхата, как уже я был з...
13    Если бы паче всякого чаяния матушка родила доч...
14              Я считался в отпуску до окончания наук.
15         В то время воспитывались мы не по нонешнему.
16    С пятилетнего возраста отдан я был на руки стр...
17    Под его надзором на двенадцатом году выучился ...
18    В это время батюшка нанял для меня француза, м...
19           Приезд его сильно не понравился Савельичу.
Name: 0, dtype: object

In [93]:
len_data = df[0].apply(len)
len_data.describe()

count    3382.000000
mean       62.975754
std        51.965178
min         1.000000
25%        26.000000
50%        49.000000
75%        84.000000
max       512.000000
Name: 0, dtype: float64

In [100]:
word_text = [nltk.word_tokenize(line) for line in text]
word_text[10]

['Нас', 'было', 'девять', 'человек', 'детей', '.']

Напишем функцию, которая будет вставлять номер телефона в предложение и возвращать измененное предложение и позицию, куда был добавлен номер.

In [102]:
def insert_phone(sentence):
    phone = generate_phone()
    position = random.choice(range(len(sentence)))
    new_sentence = sentence[:]
    new_sentence.insert(position, phone)
    return new_sentence, position

insert_phone(word_text[10])

(['Нас', 'было', 'девять', 'человек', '+7-756-815-1673', 'детей', '.'], 4)

После того, как у вас получился набор предложений с номерами промаркируйте данные. Они должны иметь такой вид:

[("Уи́льям", 0), ("Шекспи́р", 0), ("—", 0), ("английский", 0), ("8(495)3451212", 1), ("поэт", 0), ("и", 0), ("драматург", 0)]

Теперь напишем функцию, которая будет вставлять номер и возвращать размеченное предложение:

In [119]:
def label_sentence(sentence):
    sentence, position = insert_phone(sentence)
    labels = [1 if pos == position else 0 for pos in range(len(sentence))]
    return list(zip(sentence, labels))

label_sentence(word_text[14])

[('Я', 0),
 ('считался', 0),
 ('в', 0),
 ('+7-(565)-141-1488', 1),
 ('отпуску', 0),
 ('до', 0),
 ('окончания', 0),
 ('наук', 0),
 ('.', 0)]

Мы можем использовать эту функцию для того, чтобы разметить все предложения в тексте:

In [120]:
labeled_text = [label_sentence(line) for line in word_text]
labeled_text[14]

[('Я', 0),
 ('считался', 0),
 ('в', 0),
 ('отпуску', 0),
 ('до', 0),
 ('окончания', 0),
 ('наук', 0),
 ('+7-211-107-1492', 1),
 ('.', 0)]

Придумайте признаки для номера и сгенерируйте датасет.

Поскольку мы вставляем номер телефона в случайное место предложения, позиция номера в предложении не будет играть роли при классификации. Поэтому я предлагаю сформировать датасет для обучения не в виде набора предложений, а в виде размеченного набора слов.
В качестве признаков можно использовать частоту знаков, составляющих слово.

In [148]:
all_words = list()
all_labels = list()
for line in labeled_text:
    words, labels = zip(*line)
    all_words.extend(words)
    all_labels.extend(labels)

В нашем наборе данных есть сильная диспропорция между количеством всех слов и количеством телефонных номеров, телефонные номера составляют всего около 7% от всех слов -- не знаю, как это скажется на обучении:

In [147]:
sum(all_labels) * 100.0 / len(all_words)

7.204021641886423

Создадим принзнак -- частота знаков в словах:

In [152]:
from collections import Counter

all_words_with_count = [Counter(word) for word in all_words]

all_words_with_count[100]

Counter({'8': 1,
         '(': 1,
         '1': 1,
         '0': 2,
         '9': 3,
         ')': 1,
         '-': 3,
         '2': 2,
         '6': 2})

### 2. Обучение классификатора [3 балла].
Обучите один или несколько классификаторов.

In [154]:
from sklearn.metrics import accuracy_score
from nltk import NaiveBayesClassifier
from sklearn.model_selection import train_test_split

In [155]:
X_train, X_test, y_train, y_test = train_test_split(all_words_with_count, all_labels, test_size=0.1)

In [156]:
clf = NaiveBayesClassifier.train(zip(X_train, y_train))

In [166]:
pred = clf.classify_many(X_test)
accuracy_score(y_test, pred)

0.9970181043663472

Как мы видим, точность достаточно высокая. Давайте посмотрим на ошибки классификации:

In [165]:
for w, p, a in zip(X_test, pred, y_test):
    if p != a:
        print(w)

Counter({')': 1})
Counter({'(': 1})
Counter({'(': 1})
Counter({')': 1})
Counter({'(': 1})
Counter({')': 1})
Counter({')': 1})
Counter({'3': 1, '0': 1})
Counter({'(': 1})
Counter({'8': 1})
Counter({'(': 1})
Counter({')': 1})
Counter({')': 1})
Counter({')': 1})


Как мы видим -- ошибочно распознались слова, которые состоят из одной скобки или нескольких цифр. Давайте добавим еще один признак -- длину слова:

In [170]:
all_words_with_len = list()
for word in all_words:
    d = Counter(word)
    d.update({'len': len(word)})
    all_words_with_len.append(d)
    
all_words[200], all_words_with_len[200]

('детей', Counter({'д': 1, 'е': 2, 'т': 1, 'й': 1, 'len': 5}))

In [171]:
X_train, X_test, y_train, y_test = train_test_split(all_words_with_len, all_labels, test_size=0.1)

In [172]:
clf = NaiveBayesClassifier.train(zip(X_train, y_train))

In [173]:
pred = clf.classify_many(X_test)
accuracy_score(y_test, pred)

0.9995740149094782

Точность предсказания существенно улучшилась! В тестовом наборе всего две ошибки:

In [174]:
for w, p, a in zip(X_test, pred, y_test):
    if p != a:
        print(w)

Counter({'len': 2, '1': 1, '9': 1})
Counter({'len': 8, 'm': 1, 'o': 1, 'n': 1, 's': 1, 'i': 1, 'e': 1, 'u': 1, 'r': 1})


### 3. Поиск номеров регулярным выражением [3 балла].
Напишите регулярное выражение для извлечения номера из текста.

In [175]:
import re

tests = [
    '8(348)6723345',
    '8(178)-741-15-09',
    '+7-448-067-9861',
    '+7-014-658-98-99',
    '+7-(902)-652-3576',
    '8(501)3663354',
    '+7-973-117-60-30',
    '+7-052-037-2567',
    '+7-(962)-700-7442'
]

phone = re.compile(r'^(\+7|8)( *|-)\(?\d{3}\)?[\d -]{7,10}$')
for test in tests:
    m = phone.match(test)
    if m:
        print(m.group(0))
    else:
        print(f"Test {test} didn't match the pattern")

8(348)6723345
8(178)-741-15-09
+7-448-067-9861
+7-014-658-98-99
+7-(902)-652-3576
8(501)3663354
+7-973-117-60-30
+7-052-037-2567
+7-(962)-700-7442


Проверим на всем тексте:

In [176]:
for word, label in zip(all_words, all_labels):
    m = phone.match(word)
    if bool(m) != bool(label):
        print(f'Failed on word: {word}')

Регулярное выражение отработало без ошибок на нашем тексте.

### 4. Оценка качества работы [2 балла].

Охарактеризуйте результаты.

Мы обучили Байесовский классификатор на выделение телефонных номеров в тексте, используя в качестве признаков частоту знаков в слове. Точность составила 0.997, что достаточно неплохо. Добавив еще один признак, нам удалось повысить точность до 0.9996

Регулярное выражение смогло распознать все номера без ошибок на нашем пример, но пример взят не из реальной жизни -- мы сами генерировали номера и точно знаем возможные форматы. Регуляргное выражение может не предусмотреть какие-то форматы, которые могут встретиться в реальной жизни, например, более короткие номера, добавочные номера и т.п.

Одно из ограничений нашего решения в том, что мы предполагаем, что номер телефона выделяется в отдельный токен. Это учловие соблюдается, если телефонные номера не содержат пробелов. В противном случае, выделение номеров лучше проводить на этапе токенизации.

Ответьте на вопросы:
* правильно ли мы генерировали данные для задачи?
* какие признаки и классификаторы будут работать лучше, если мы знаем, как были получены данные.

Мы генерировали телефонные номера по шаблонам, которые сами же и придумали. Лучше было бы взять примеры текстов, содержащие телефонные номера и предварительно их разметить.

Регулярные выражения хорошо работают, когда мы знаем все возможные форматы данных. Байесовский классификатор будет лучше работать в условиях реальных данных, но для него нужна правильная токенизация.

## Сдача домашнего задания

Дедлайн сдачи домашнего задания:  23:59 28.04.2018. 

Результаты домашнего задания должны быть оформлены в виде отчета в jupyter notebook.
Нормальный отчёт должен включать в себя:
* Краткую постановку задачи и формулировку задания
* Описание минимума необходимой теории и/или описание используемых инструментов 
* Подробный пошаговый рассказ о проделанной работе
* **Аккуратно** оформленные результаты
* Подробные и внятные ответы на все заданные вопросы 
* Внятные выводы – не стоит относится к домашнему заданию как к последовательности сугубо технических шагов, а стоит относится скорее как к небольшому практическому исследованию, у которого есть своя цель и свое назначение.


Сдача отчетов осуществляется по email: login-const@mail.ru, заголовок письма – NLP-HSE-HW1-ИмяФамилия. 
